### DISCUSSION TITLE
サブビューを生成する処理は Computed-property よりメソッドを好む。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#8] [Ideas]
## Overview

SwiftUI において `body` 関数が大きくなりすぎた場合、`private` な Computed-property またはメソッドに分割すると、View 内に閉じた共通化や可読性の向上に繋がる場合がある。

計算コストがあることが明白で、かつあとから引数の導入もしやすいメソッドを好むとよい。

```swift
struct ContentView: View {
    var sidebarTitle = "title"
    
    var body: some View {
        VStack {
            // 🤔 Use property. (No computational cost?)
            header

            HStack {
                // 💡Use method.
                content()
                sidebar(title: sidebarTitle) // ✅ Easy to introduce arguments later.
            }
        }
    }

    // 🤔 If arguments are to be introduced, they must be rewritten into functions.
    private var header: some View {
        Text("Menu")
    }

    private func content() -> some View {
        Text("Content")
    }

    private func sidebar(title: String) -> some View {
        Text("Sidebar: \(title)")
    }
}
```

## Discussion

Computed-property の方がわずかに記述量が少なく、スッキリして見えるため最初は Computed-property で記述したくなるが、あとから引数を導入したくなることは開発時において決して少なくないため、最初からメソッドで記述しておくほうがベターと思われる。

また、変数参照ではなく”呼び出し”であることも呼び出し元のコードを見るだけで明白になるのも利点。

## Why not?

ただし、好みによって分かれる部分でもあるため、チームで開発する際は最初に方針を決めておくと、コーディングスタイルが統一され、PR における自転車置き場の議論を避けることにつながる。

## Note

複数の View から共通的に利用可能であれば、別の View への切り出しも検討すること。

**Comment by uhooi** [comment]
`Text()` 1行のようにシンプルな View はメソッドで返すのでいいと思いますが、複雑な View は使い回さなくても View に準拠した struct に切り出すのがいいと思いました！
例がパッとは出てこないのですが、struct を使わないと実現できない View もある気がします。

**Reply by uhooi** [reply]
最後の「Note」に対するコメントですー！

**Comment by yuukiw00w** [comment]
シンプルなViewはbody内でlet定義する方が記述量が少なくbody内以外で使わないことも明らかになるので読みやすくなると思います
複雑なViewの場合はpreviewで見たくなる程度の量になってくると思われるため、structに切り出す方がいいと思いました
そのViewでしか使わない場合は同じファイル内でprivate structで切り出しておくと可視性制御もできると思います

例
```swift
struct ContentView: View {
    var sidebarTitle = "title"
    
    var body: some View {
        let header = Text("Menu")
        let content = Text("Content")
        let sidebar = Text("Sidebar: \(sidebarTitle)")

        VStack {
            header

            HStack {
                content
                sidebar
            }
        }
    }
}
```

SwiftUIのWWDCサンプルアプリの例
https://github.com/cntrump/SwiftUIAccessibilityExamples/blob/5966af1987042fe96cfdc8bcdcddaf244d19166e/SwiftUIAccessibilityExamples/Shared/AccessibilityExamples.swift#L96
https://github.com/cntrump/SwiftUIAccessibilityExamples/blob/master/SwiftUIAccessibilityExamples/Shared/CustomControlsExample.swift#L144

なおWWDCサンプルアプリでは結構ベタッとbody内に書いてたりcomputed propertyだったり色々な模様
